### PR TITLE
Multi model profile top level

### PR DESCRIPTION
**Pull request by nv-braf**
Changes to Analyzer and ModelManager to support multi-model.

### CODE REVIEW

File: model_analyzer/config/input/config_command_profile.py [hunk 1]
```diff
@@ -483,6 +483,16 @@ def _add_run_search_configs(self):
...
+        self._add_config(
+            ConfigField(
+                'run_config_profile_models_concurrently_enable',
+                flags=['--run-config-profile-models-concurrently-enable'],
+                field_type=ConfigPrimitive(bool),
+                parser_args={'action': 'store_true'},
+                default_value=
+                DEFAULT_RUN_CONFIG_PROFILE_MODELS_CONCURRENTLY_ENABLE,
+                description=
+                "Enable the profiling of multiple models concurrently."))
```
Comment (tgerdesnv) [inline model_analyzer/config/input/config_command_profile.py:495]: We only support all or nothing, right? I think the description should say "profiling of all supplied models concurrently"

File: model_analyzer/config/input/config_defaults.py [hunk 2]
```diff
@@ -41,8 +41,6 @@
 DEFAULT_RUN_CONFIG_MAX_CONCURRENCY = 1024
 DEFAULT_RUN_CONFIG_MAX_INSTANCE_COUNT = 5
 DEFAULT_RUN_CONFIG_SEARCH_DISABLE = False
-DEFAULT_RUN_CONFIG_MAX_PREFERRED_BATCH_SIZE = 16
```
Comment (tgerdesnv) [inline model_analyzer/config/input/config_defaults.py:44]: Why is this showing up? Are we not properly rebased?

Comment (nv-braf) [reply model_analyzer/config/input/config_defaults.py]: Yes. I've reset, squashed and rebased properly.

**Review by tgerdesnv — APPROVED** [review]

File: model_analyzer/config/input/config_command_profile.py [hunk 3]
```diff
@@ -491,7 +491,7 @@ def _add_run_search_configs(self):
                 default_value=
                 DEFAULT_RUN_CONFIG_PROFILE_MODELS_CONCURRENTLY_ENABLE,
                 description=
-                "Enable the profiling of multiple models concurrently."))
+                "Enable the profiling of all supplied models concurrently."))
```
Comment (matthewkotila) [inline model_analyzer/config/input/config_command_profile.py:494]: copyright year should be `2021-2022`

File: model_analyzer/model_manager.py [hunk 4]
```diff
@@ -53,25 +53,26 @@ def __init__(self, config, client, server, metrics_manager, result_manager,
         self._result_manager = result_manager
         self._state_manager = state_manager
 
-    def run_model(self, model):
+    def run_models(self, models):
```
Comment (matthewkotila) [inline model_analyzer/model_manager.py:56]: copyright year should be `2021-2022`

File: tests/test_model_manager.py [hunk 5]
```diff
@@ -674,7 +674,7 @@ def _test_model_manager(self, yaml_content, expected_ranges):
                                      MagicMock(), MagicMock(), metrics_manager,
                                      MagicMock(), state_manager)
 
-        model_manager.run_model(config.profile_models[0])
+        model_manager.run_models(config.profile_models[0])
```
Comment (matthewkotila) [inline tests/test_model_manager.py:677]: copyright year should be `2021-2022`

File: model_analyzer/analyzer.py [hunk 6]
```diff
@@ -109,14 +109,22 @@ def profile(self, client, gpus):
...
 
-        # Profile each model, save state after each
-        for model in self._config.profile_models:
-            if self._state_manager.exiting():
-                break
+        # Profile all models concurrently
+        if self._config.run_config_profile_models_concurrently_enable:
             try:
-                self._model_manager.run_model(model=model)
+                self._model_manager.run_models(
```
Comment (matthewkotila) [inline model_analyzer/analyzer.py:115]: why is this in a try block?

File: model_analyzer/model_manager.py [hunk 7]
```diff
@@ -53,25 +53,26 @@ def __init__(self, config, client, server, metrics_manager, result_manager,
...
         Generates configs, runs inferences, gets
-        measurements for a single model
-
+        measurements for a list of models
         Parameters
         ----------
-        model : ConfigModelProfileSpec
-            The model being run
+        models : List of ConfigModelProfileSpec
+            The list models being run
```
Comment (matthewkotila) [inline model_analyzer/model_manager.py:63]: small nit:
> The list _of_ models being run

**Review by matthewkotila — CHANGES_REQUESTED** [review]

File: model_analyzer/analyzer.py [hunk 8]
```diff
@@ -109,14 +109,22 @@ def profile(self, client, gpus):
             self._metrics_manager.profile_server()
             self._server.stop()
 
-        # Profile each model, save state after each
-        for model in self._config.profile_models:
-            if self._state_manager.exiting():
-                break
+        # Profile all models concurrently
```
[the remark this answers is not in the document]
Comment (nv-braf) [reply model_analyzer/analyzer.py:112]: Fixed

File: model_analyzer/model_manager.py (hunk 4 above)
Comment (nv-braf) [reply model_analyzer/model_manager.py:56]: Fixed

File: model_analyzer/config/input/config_command_profile.py (hunk 3 above)
Comment (nv-braf) [reply model_analyzer/config/input/config_command_profile.py:494]: Fixed

File: tests/test_model_manager.py (hunk 5 above)
Comment (nv-braf) [reply tests/test_model_manager.py:677]: Fixed

File: model_analyzer/analyzer.py (hunk 6 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:115]: I don't know? I just copied what already existed

File: model_analyzer/model_manager.py (hunk 7 above)
Comment (nv-braf) [reply model_analyzer/model_manager.py:63]: Fixed

**Review by matthewkotila — APPROVED** [review]

**Pull request merged**